RFS/src/main/java/com/rfs/worker/IndexStep.java

```diff
@@ -0,0 +1,475 @@
+                 * it means someone else has either marked it as completed or failed, and either is fine.
+                 * 
+                 * If we fail to migrate it, we attempt to increment the attempt count.  It's fine if the increment
+                 * fails because we guarantee that we'll attempt the work at least N times, not exactly N times.
+                 */
+                if (workItem.numAttempts > CmsEntry.IndexWorkItem.ATTEMPTS_SOFT_LIMIT) {
+                    logger.info("Index Work Item " + workItem.name + " has exceeded the maximum number of attempts; marking it as failed...");
+                    CmsEntry.IndexWorkItem updatedEntry = new CmsEntry.IndexWorkItem(
+                        workItem.name,
+                        CmsEntry.IndexWorkItemStatus.FAILED,
+                        workItem.numAttempts,
+                        workItem.numShards
+                    );
+
+                    members.cmsClient.updateIndexWorkItem(updatedEntry, workItem).ifPresentOrElse(
+                        value -> logger.info("Index Work Item " + workItem.name + " marked as failed"),
+                        () ->logger.info("Unable to mark Index Work Item " + workItem.name + " as failed")
+                    );
+                    continue;
+                }
+
+                try {
+                    logger.info("Migrating index: " + workItem.name);
+                    IndexMetadata.Data indexMetadata = members.metadataFactory.fromRepo(members.snapshotName, workItem.name);
+
+                    ObjectNode root = indexMetadata.toObjectNode();
+                    ObjectNode transformedRoot = members.transformer.transformIndexMetadata(root);
+
+                    IndexMetadataData_OS_2_11 indexMetadataOS211 = new IndexMetadataData_OS_2_11(transformedRoot, indexMetadata.getId(), workItem.name);
+                    members.indexCreator.create(workItem.name, indexMetadataOS211).ifPresentOrElse(
+                        value -> logger.info("Index " + workItem.name + " created successfully"),
+                        () -> logger.info("Index " + workItem.name + " already existed; no work required")
+                    );
+
+                    logger.info("Forcefully updating the Index Work Item to indicate it has been completed...");
+                    CmsEntry.IndexWorkItem updatedEntry = new CmsEntry.IndexWorkItem(
+                        workItem.name,
+                        CmsEntry.IndexWorkItemStatus.COMPLETED,
+                        workItem.numAttempts,
+                        workItem.numShards
+                    );
+                    members.cmsClient.updateIndexWorkItemForceful(updatedEntry);
+                    logger.info("Index Work Item updated");
+                } catch (Exception e) {
+                    logger.info("Failed to migrate index: " + workItem.name, e);
+                    logger.info("Updating the Index Work Item with incremented attempt count...");
```

Comment:
these should be at a higher log level.  
We should think about how permanent failures are propagated back to the user/console too.

Reply:
Per discussion - this isn't a permanent failure.